src/compiler/rust.rs

```diff
@@ -0,0 +1,761 @@
+use log::LogLevel::Trace;
+use mock_command::{CommandCreatorSync, RunCommand};
+use sha1;
+use std::borrow::Cow;
+use std::collections::{HashMap, HashSet};
+use std::fs::File;
+use std::io::Read;
+use std::iter::FromIterator;
+use std::path::{Path, PathBuf};
+use std::process;
+use std::time::Instant;
+use tempdir::TempDir;
+use util::{fmt_duration_as_secs, sha1_digest};
+
+use errors::*;
+
+/// A unit struct on which to hang a `Compiler` impl.
+#[derive(Debug, Clone)]
+pub struct Rust;
+
+/// A struct on which to hang a `CompilerHasher` impl.
+#[derive(Debug, Clone)]
+pub struct RustHasher {
+    /// The full commandline.
+    arguments: Vec<String>,
+    /// The commandline without any --emit or --out-dir arguments.
+    filtered_arguments: Vec<String>,
+    /// The location of compiler outputs.
+    output_dir: String,
+    /// Paths to extern crates used in the compile.
+    externs: Vec<String>,
+    /// The crate name passed to --crate-name.
+    crate_name: String,
+    /// If dependency info is being emitted, the name of the dep info file.
+    dep_info: Option<String>,
+}
+
+/// A struct on which to hang a `Compilation` impl.
+#[derive(Debug, Clone)]
+pub struct RustCompilation {
+    /// The full commandline.
+    arguments: Vec<String>,
+    /// The compiler outputs.
+    outputs: HashMap<String, String>,
+    /// The crate name being compiled.
+    crate_name: String,
+}
+
+/// Arguments that take a value.
+const ARGS_WITH_VALUE: &'static [&'static str] = &[
+    // These are taken from https://github.com/rust-lang/rust/blob/b671c32ddc8c36d50866428d83b7716233356721/src/librustc/session/config.rs#L1186
+    "--cfg",
+    "-L",
+    "-l",
+    "--crate-type",
+    "--crate-name",
+    "--emit",
+    "--print",
+    "-o",
+    "--out-dir",
+    "--explain",
+    "--target",
+    "-W", "--warn",
+    "-A", "--allow",
+    "-D", "--deny",
+    "-F", "--forbid",
+    "--cap-lints",
+    "-C", "--codegen",
+    "--extern",
+    "--sysroot",
+    "-Z",
+    "--error-format",
+    "--color",
+    "--pretty",
+    "--unpretty",
+];
+
+/// Emit types that we will cache.
+const ALLOWED_EMIT: &'static [&'static str] = &["link", "dep-info"];
+
+/// Version number for cache key.
+const CACHE_VERSION: &'static [u8] = b"1";
+
+/// Parse the value passed to a commandline argument `arg`, either after
+/// an '=' in `arg` or in the next argument from `it`.
+fn arg_val<'a, 'b, T>(arg: &'a str, it: &'b mut T) -> Option<&'a str>
+    where T: Iterator<Item=&'a String>,
+{
+    if let Some(i) = arg.find('=') {
+        Some(&arg[i+1..])
+    } else {
+        it.next().map(|v| v.as_str())
+    }
+}
+
+/// Return true if `arg` is in the set of arguments `set`.
+fn arg_in(arg: &str, set: &HashSet<&str>) -> bool
+{
+    set.contains(arg) || set.iter().any(|a| arg.starts_with(a))
+}
+
+/// Calculate the SHA-1 digest of each file in `files` on background threads
+/// in `pool`.
+fn hash_all(files: Vec<String>, pool: &CpuPool) -> SFuture<Vec<String>>
```

Comment:
I sort the list of source files from the dep-info, but not the list of externs, they're input to the hash in the order they're listed on the commandline.

If cargo produces nondeterministic commandlines we've got a bigger problem, since we're feeding the entire commandline as hash input! Can we verify that assumption?